4 단계 - 자동차 경주(우승자)

Jinni01 opened the pull request:
커밋을 요약이라 생각하고 올리긴 했는데, 보기에 불편하실 수도 있을 것 같아 앞으론 요약해서 올리도록 하겠습니다. 

- RacingCar 클래스 이름 프로퍼티 추가
- 이름 프로퍼티 글자 제한 
- RacingCar 클래스 변경에 따른 Factory, Container 클래스 내용 변경 ->  RaceSetting, RacingCarSetting 정보 받아 생성하도록 함
- 우승자 출력 로직 추가 
- 우승자 출력 로직에 거리 정보 추가
- 추가된 이름 프로퍼티에 대한 모든 테스트 케이스 검증 단언문 추가

Jinni01 committed:
1 단계 - 코틀린 기초 (#4)

* feat: java to kotlin

* feat: Person class kotest 이용해 테스트 케이스 작성

---------

Co-authored-by: suj0324 <suj0324@midasit.com>

Jinni01 committed:
2 단계 - 문자열 계산기 (#5)

* feat: Step1 feedback 반영

-  data class 선언시 제공되는 메서드 활용 시도

* feat: 계산기 클래스 추가 및 입력값 분해 테스트 케이스 작성

* feat: 공백 문자열 입력 테스트 케이스 작성

* feat: null 입력 테스트 케이스 작성

* feat: 더하기 연산 테스트 케이스 작성

* feat: 빼기 연산 테스트 케이스 작성

* feat: 곱하기 연산 테스트 케이스 작성

* feat: 나누기 연산 테스트 케이스 작성

* feat: 목록에 없는 연산 기호 사용 테스트 케이스 작성

* feat: 사칙 연산 테스트 케이스 작성

* refactor: Step2 피드백 반영 및 개선

- StringBasedCalculatorV2 추가(개선판, 머지시 기존 클래스 교체)
- 계산기 클래스 인스턴스에서 유틸클래스로 변경(정적메소드 사용)
- 나누기 연산 0 테스트 케이스 추가
- 디버깅용 프린트구문 제거
- if문 예외 체크 require로 변경

* refactor: Step2 피드백 반영 및 개선

- 리스트 생성자 인자 방어적 복사
- 에러메시지 문자열 타입 더하기 연산 제거
- dto -> 용도에 맞지 않는 클래스명 변경
- 불필요 지역변수 할당 제거
- StringBasedCalculator 싱글턴 객체로 변경
- 테스트 케이스 추가: 잘못된 수식
- 테스트 케이스 추가: 커스텀 구분자 사용
- 테스트 케이스 추가: 피연산자 위치에 다른 값 입력
- 내부 enum 캐싱하도록 변경
- 사용되지 않는 클래스 Deprecated 처리 (다음커밋시 제거)

---------

Co-authored-by: suj0324 <suj0324@midasit.com>

Jinni01 committed:
3 단계 - 자동차 경주 (#11)

* feat: Step2 feedback 반영 및 개선

- enum 리플렉션반환 코드 제거
- 코틀린 collection api 활용(zip, fold)

* feat: Step2 feedback 반영 및 개선

- 불필요 주석 제거

* feat: 레이싱카 객체 이동 테스트 케이스 작성

* feat: RacingCarPosition 도메인 분리

* feat: racingCar attemptMove 메서드 추가

* feat: RaceStrategy 인터페이스 및 RandomStrategy 싱글턴 객체 추가

* feat: RaceContainer 및 RacingCarFactory 클래스 추가

- ErrorMessage 클래스 추가

* feat: RacingCarFactory 싱글턴으로 변경

* feat: Race클래스 추가, 스텝/전체 기록 남기는 로직 추가

* feat: InputView/ResultView 클래스 추가, main 함수 작성

* feat : RaceHistoryDto data class 변경

* feat : code lint

* refactor: 프로젝트 패키지 구조 스텝별 분리 및 개선

- 각각의 스텝을 루트로 형상 가져가도록 수정

* feat: 자동차 경주 요구 사항 정의서 작성

* feat: RacingCarPosition value class화, 테스트 케이스 추가

- 빌드 에러 발생시키는 미완성 클래스들 제거

* feat: RacingCarPosition 뺴기 연산 테스트 케이스 추가

* feat: RacingCar 클래스 리팩토링, 테스트 케이스 추가

- position 주입 가능하고, 참조 변경할 수 있도록 수정
- 바뀐 전략에 맞춘 attemptMove 메서드 작성

* feat: RandomStrategy 클래스 리팩토링

- 바뀐 인터페이스 정책에 따라 내부 구현 수정

* feat: RacingCar 클래스 수정

- 내부에서 정책을 소유하는 것이 아닌, 이동시 메서드에서 인자로 받도록 수정
- 연관된 테스트 케이스 코드 수정

* feat: RacingCarFactory 클래스 리팩토링

- 레이싱카 리스트 생성 메서드에서 정책을 주입받지 않도록 수정(레이싱카 이동 메서드 변경 관련 반영)
- 생성될 레이싱카들의 초기 위치 함수 파라미터 추가

* feat: RacingCarContainer 클래스 리팩토링

- 생성될 레이싱카들의 초기 위치 함수 파라미터 추가
- StepHistoryDto 도메인 객체에 의존하지 않고 정수 결과값 리스트 반환하도록 수정
- List<RacingCar> 타입의 확장 함수 추가

* feat: RacingCarContainer 클래스 테스트 케이스 작성

- inspector 활용

* feat: Race 클래스 리팩토링, 테스트 케이스 작성

- start 메서드 로직 수정, 코틀린 문법 활용

* feat: 체크리스트 반영

* feat: GameController / GameService 싱글턴 클래스 추가

* feat: 입출력 관련 작업 추가, 각 동작 연결

* refactor: request change 내용 반영 및 개선

- 테스트 대상 클래스별로 테스트클래스 작성(분리)
- 실패 테스트 케이스에 대한 로직 수정
- 출력문 로직 개선(결과 문자열 빌드)
- Factory/Container 객체에서 레이싱카 설정 정보를 통해 차별화할 수 있도록함
- 불필요 전위 반환 로직 수정
- named argument 활용
- Race 클래스 제거 및 RaceGameService 로직 통합

* refactor: request change 내용 반영 및 개선

- requirements.md 변경 내용 반영

---------

Co-authored-by: suj0324 <suj0324@midasit.com>

Commit:
feat: Step3 피드백 반영 및 개선

- RacingGameService start 메서드 테스트 케이스 추가
- String DSL 활용
- pure for문 코틀린 문법 활용해서 변경
- kotest 종속성 버전업

Commit:
feat: RacingCar 이름 포함 생성 테스트 케이스 작성

Commit:
feat: RacingCar 이름 포함 생성 테스트 케이스 작성

Commit:
feat: RacingCar 클래스 이름 프로퍼티 추가

Commit:
feat: RacingCar 클래스 5글자를 넘는 이름 포함 생성 테스트 케이스 추가

Commit:
feat: RacingCar 변경에 따른 참조 소스 및 테스트 케이스 동반 수정

Commit:
feat: 입력값 포맷 변경

- 관련 내부 로직 변경
- 새로운 입력 로직 추가

Commit:
feat: ResultView 우승자 출력 로직 추가

Commit:
feat: ResultView 우승자 출력 문구 포맷 수정

lhs0610 commented on src/main/kotlin/racingcar/domain/RacingCar.kt:
```diff
@@ -0,0 +1,32 @@
+import racingcar.strategy.RaceStrategy
+
+class RacingCar(
+    val racerName: String,
```
이름도 일급 객체로 포장(Wrapping)한다면 이름 값 5글자 체크하는 유효성도 제거할 수 있지 않을까요?

lhs0610 commented on src/main/kotlin/racingcar/domain/RacingCarContainer.kt:
```diff
@@ -0,0 +1,28 @@
+    }
+}
+
+fun List<RacingCar>.getAllDistance(): List<Int> {
```
현재 사용중인 메서드가 아닌 것 같은데, 사용하지 않는 함수는 제거해주세요.

Jinni01 replied:
확장함수를 처음 사용해본 예시라 기념으로 잔존해두었는데 제거하도록 하겠습니다.

항상 지울때 가장 망설이게 되네요 ㅎㅎㅎ

lhs0610 commented on src/main/kotlin/racingcar/domain/RacingCarPosition.kt:
```diff
@@ -0,0 +1,19 @@
+package racingcar.domain
+
+@JvmInline
+value class RacingCarPosition(val distance: Int) {
```
객체명은 자동차위치, 값은 거리면 뭔가 매칭이 애매하게 되는 것 같아요 좀 더 단순하게 클래스 시그니처를 보여주면 어떨까요?

Jinni01 replied:
나중에y축이 추가되면 어쩌지라는 상황을 가정해서 처음에 지었던 것 같은데, 동일하게 이름을 가져가는 것이 덜 헷갈리고 직관적인 것 같습니다. 수정해보겠습니다

lhs0610 commented on src/main/kotlin/racingcar/view/InputView.kt:
```diff
@@ -0,0 +1,82 @@
+        return StringSplitter.split(input)
+    }
+
+    private fun validateEntryInput(input: String): Boolean {
```
UI Layer에서 domain의 유효성 검증로직에 대한 의존성을 다 같이 가지고 있는것 같은데, 어떻게 분리할 수 없을까요? 
스프링을 고민해보면 좋을 것 같아요.

lhs0610 commented on src/main/kotlin/stringcalculator/component/StringBasedCalculator.kt:
```diff
@@ -0,0 +1,52 @@
+        val operators = ArrayDeque<String>()
+        val operands = ArrayDeque<Int>()
+
+        for ((index, value) in split.withIndex()) {
+            if (isOdd(index)) {
+                val operator = getExistOperator(value)
+                operators.add(operator)
+                continue
+            }
+
+            val operand = convertStringToIntCheckNumeric(value)
+            operands.add(operand)
+        }
```
반복문안에 분기문이 있다보니 indent가 3이네요. 줄여볼 수 있을 것 같아요.

Jinni01 replied:
요건 문자열 계산기 스텝에 있던 내용이군요. 별도로 수정해서 올리도록 하겠습니다

lhs0610 commented on src/main/kotlin/racingcar/strategy/RandomStrategy.kt:
```diff
@@ -0,0 +1,14 @@
+package racingcar.strategy
+
+object RandomStrategy: RaceStrategy {
```
범위값 테스트 정도는 진행해볼 수 있을 것 같아요 reapeat해서 범위를 벗어나지 않는지 말이죠. 완벽하진 않더라도
어느정도의 capacity는 안전성을 채울 수 있으니까요

Jinni01 replied:
수행 결과가 예상 범위 내인지 테스트하는 단언문을 작성해볼 수 있도록 해보겠습니다.

lhs0610 commented on src/main/kotlin/racingcar/model/RaceSetting.kt:
```diff
@@ -0,0 +1,11 @@
+data class RaceSetting(val racingcarSettings: List<RacingCarSetting>, val totalStep: Int) {
+    companion object {
+        fun from(userInput: UserInput): RaceSetting {
+            val racingcarSettings = userInput.entries.map { RacingCarSetting(it) }
```
카멜케이스라면 racingCarSettings여야 할 것 같네요

Jinni01 replied:
하나의 단어로 생각해서 묶었었는데 전체적으로 수정하도록 하겠습니다.

lhs0610 commented on src/main/kotlin/racingcar/component/RacingCarFactory.kt:
```diff
@@ -0,0 +1,11 @@
+
+object RacingCarFactory {
+    fun generateRacingCars(settings: List<RacingCarSetting>): List<RacingCar> {
+        return settings.map { RacingCar(it.racerName, RacingCarPosition(it.initPosition)) }.toList()
```
로직만 봐선 자동차 일급 컬렉션에서 팩토리 함수로 가져가도 상관 없을 것 같네요.

Jinni01 replied:
팩토리 클래스를 분리했었는데 정적 팩토리 함수에서 변환해도 될 것 같네요!

lhs0610 commented on src/main/kotlin/racingcar/view/ResultView.kt:
```diff
@@ -0,0 +1,38 @@
+
+    fun printRacingGameResult(raceHistory: RaceHistory) {
+
+        raceHistory.stepHistories.forEach() {
```
괄호문은 없어도 될 것 같아요.

lhs0610 commented on src/main/kotlin/racingcar/view/ResultView.kt:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    private fun printWinners(lastStepHistory: StepHistory) {
+        val reverseSortedGroups = lastStepHistory.stepRacingCarInfos.groupBy(
+            { it.distance }, { it.racerName }
+        ).toSortedMap(compareBy<Int> { it }.reversed())
```
출력계층에서 우승자 판별 로직을 구현하고 있는데, 적절한가요?

Jinni01 replied:
지금 보니 단순한 정보의 출력이 아닌 우승자 판별까지 수행하는건 맡은 역할에 적절하지 않은 것 같습니다. 분리해서 본 역할만 가질 수 있도록 해보겠습니다.

lhs0610 reviewed (changes requested):
의진님 전체적으로 3단계에서 구조가 잡혀있어서 4단계에서는 크게 피드백 드릴 부분이 많이 없는 것 같아요. 근데 klintcheck를 계속 안하시네요 🤔

피드백 반영 후 다시 리뷰요청 해주세요.

Commit:
refactor: request change 내용 반영 및 개선

- racingcar -> racingCar camelcase에 맞게 이름 수정
- RacingCarPosition -> RacingCarDistance 클래스명 수정
- foreach 구문 불필요 괄호 제거
- RaceHistory 클래스에 우승자 그룹 조회 함수 추가(출력계층 로직 이전)
- RaceFactory 클래스 제거, RacingCarContainer 클래스 팩토리 메서드에 로직 병합
- RandomStrategy 테스트 클래스 및 케이스 추가
- 미사용 메서드 제거
- RacingCar 클래스 of 정적 팩토리 메서드 추가
- InputView 입력값 검증 및 변환 로직 argumentResolver에 위임

Commit:
refactor: request change 내용 반영 및 개선

- TotalStepArgumentResolver 테스트 클래스 및 케이스 추가
- EntryArgumentResolver 테스트 클래스 및 케이스 추가
- InputArgumentResolvers 테스트 클래스 및 케이스 추가
- 요구사항정의서 수정 반영

lhs0610 commented on src/main/kotlin/racingcar/config/Configuration.kt:
```diff
@@ -0,0 +1,15 @@
+
+object Configuration {
+
+    fun init() {
```
- 명시적으로 초기화 메서드를 만들 이유가 따로 있나요? 
- 각 리졸버들도 stateless라서 무상태 객체로 보이는데 객체 생성을 해주지 않아도 상관없을 것 같네요

lhs0610 commented on src/main/kotlin/racingcar/converter/EntryArgumentResolver.kt:
```diff
@@ -0,0 +1,38 @@
+    override fun validate(input: String): Boolean {
+        val split = StringSplitter.split(input)
+        if (split.any { it == BLANK }) {
+            println(Message.INVALID_INPUT_FORTMAT_CANNOT_SPLIT.message)
```
비즈니스 로직에 출력문이 없도록 해주세용

Jinni01 replied:
기존 코드를 재활용하다보니 남아있는 부분이 있었네요. 제거할 수 있도록 하겠습니다.

lhs0610 commented on src/main/kotlin/racingcar/converter/EntryArgumentResolver.kt:
```diff
@@ -0,0 +1,38 @@
+import racingcar.domain.StringSplitter
+import racingcar.model.EntryInput
+
+class EntryArgumentResolver : InputArgumentResolver<EntryInput> {
```
내부는 자동차 참가자들의 이름들에 대한 유효성 검증으로 보이는데, 하나의 유효성 검증 로직에 많은 검증이 다 들어가다보니 15라인을 넘어가고 한 메서드가 하나의 기능을 한다고 하기엔 너무 많은 기능이 들어가네요. 

그리고 검증로직도 좀 더 의미있게 줄일 수 있지 않을까요.

lhs0610 commented on src/main/kotlin/racingcar/converter/InputArgumentResolvers.kt:
```diff
@@ -0,0 +1,30 @@
+        return type.cast(resolved)
+    }
+
+    fun <T : Any> validate(type: KClass<T>, value: String): Boolean {
```
resolver 가 검증기의 역할까지 같이 가져가고 있는 것 같아요 🤔

lhs0610 commented on src/main/kotlin/racingcar/converter/InputArgumentResolvers.kt:
```diff
@@ -0,0 +1,30 @@
+object InputArgumentResolvers {
+    private val resolverMap = mutableMapOf<KClass<*>, InputArgumentResolver<*>>()
+
+    fun <T : Any> resolve(type: KClass<T>, value: String): T {
```
무슨 마음으로 이걸 작성하셨는지 짐작이 갑니다. 
잘하셨습니다 👍

lhs0610 commented on src/main/kotlin/racingcar/domain/RacingCar.kt:
```diff
@@ -1,19 +1,18 @@
-    init {
-        require(racerName.length <=5 ) { Message.INVALID_INPUT_VALUE_CANNOT_HIGHER_THEN_FIVE.message }
+    fun getDistance(): Int {
+        return distance.distance
```
클래스명과 변수명이 같을 경우 이렇게 혼동을 주기 쉬운 코드가 될 수 있어요.

Jinni01 replied:
잘 구분될 수 있도록 좋은 이름을 지어보도록 하겠습니다

lhs0610 commented on src/main/kotlin/racingcar/domain/RacingCar.kt:
```diff
@@ -1,19 +1,18 @@
-class RacingCar(
-    val racerName: String,
-    private var position: RacingCarPosition
+class RacingCar private constructor(
```
private한 생성자로 둬야 할 이유가 있을까요?

Jinni01 replied:
정적 팩토리만 열어두어도 되지 않을까 생각했는데

별도 생성자 역할로 제공하고, 값 타입을 인자로 받는 기존 생성자도 제공해도 될 것 같습니다.

lhs0610 commented on src/test/kotlin/racingcar/EntryArgumentResolverTest.kt:
```diff
@@ -0,0 +1,52 @@
+
+    context("EntryArgumentResolver test") {
+        expect("validate test - valid format") {
+            val resolver = EntryArgumentResolver()
```
이 Resolver는 context레벨에서 초기화되도 될 것 같지 않나요?

Jinni01 replied:
별도의 상태 조작이 없어 한 번만 초기화해서 사용해도 되었을 것 같습니다. 

변수를 따로 빼는게 변경 예정 사항이었겠지만 무상태이기 때문에 object class로 변경하고 참조 변경하도록 하겠습니다.

lhs0610 commented on src/main/kotlin/racingcar/model/RaceHistory.kt:
```diff
@@ -1,3 +1,14 @@
+import kotlin.collections.MutableMap.MutableEntry
+
+data class RaceHistory(val stepHistories: List<StepHistory>) {
+    fun getWinnerGroupEntry(): MutableEntry<Int, List<String>> {
```
해당 메서드에 대한 테스트도 있으면 좋을 것 같네요.

lhs0610 commented on src/main/kotlin/racingcar/model/RaceSetting.kt:
```diff
@@ -1,11 +1,11 @@
-data class RaceSetting(val racingcarSettings: List<RacingCarSetting>, val totalStep: Int) {
+data class RaceSetting(val racingCarSettings: List<RacingCarSetting>, val totalStep: Int) {
     companion object {
         fun from(userInput: UserInput): RaceSetting {
```
해당 팩토리 함수에 대한 테스트도 있으면 좋을 것 같네요.

lhs0610 reviewed (changes requested):
의진님 4단계 미션 잘 진행해주셨어요.
초반과 다르게 이제 객체 분리도 잘 되면서 여러 코틀린 기능도 쓰려고 노력하시니 보기 좋네요 👍 
차세대 리더로 오래오래 있어주시죠 🤣

사실 이제 큰 문제는 없어 바로 merge해도 괜찮을 것 같은데, 몇 가지 피드백 드렸으니 확인 후 리뷰요청 해주세요.

Commit:
refactor: request change 내용 반영 및 개선

- InputArgumentResolver 구현체들 싱글턴 클래스로 변경
- Message enum 오타 수정
- RaceSetting 팩토리 메서드 테스트 케이스 추가
- RaceHistory 우승자 반환 로직 테스트 케이스 추가
- 검증 로직 validator로 별도 분리

Commit:
refactor: request change 내용 반영 및 개선

- validator / resolver 분리에 따른 테스트 케이스 분리
- 구현체들 기준 context scope 지정

Jinni01 commented:
ktlint를 초반에 습관화가 안되어있어서 제대로 사용하지 않았는데, 커밋체크에 자동으로 되도록 걸어두었습니다. 꼼꼼히 체크할 수 있도록 하겠습니다 하하

Commit:
feat: test 동작 실패 요인 제거

Commit:
feat: klint 규격 맞춤

lhs0610 commented on src/main/kotlin/racingcar/validator/EntryArgumentValidator.kt:
```diff
@@ -0,0 +1,36 @@
+object EntryArgumentValidator : InputArgumentValidator {
+    private const val BLANK = ""
+
+    override fun validate(input: String): Boolean {
```
Spring에서는 검증기가 실패했을때 단순 boolean 타입변환이 아니라 좀 더 자세한 정보를 반환하는데 다음에는 이 부분도 고려해봐도 좋을 것 같아요 👍

Jinni01 replied:
저도 객체나 상태 열거 타입을 반환해서 ui레이어에서도 원인을 알 수 있게 하면 괜찮겠다는 생각을 했었습니다.

검증기를 만들 때 참고할 수 있도록 해 보겠습니다 !!

lhs0610 commented on src/main/kotlin/racingcar/validator/EntryArgumentValidator.kt:
```diff
@@ -0,0 +1,36 @@
+        return split.map { it.length }.any { value -> value <= 5 }
+    }
+
+    private fun checkDuplicate(split: List<String>): Boolean {
```
중복을 허용하지 않는 구조라면List가 아니라 Set을 쓰면 어떨까싶네요.

Jinni01 replied:
인자에서 set타입의 컬렉션을 받게 되면 내부에서 중복된 요소를 찾을 수 있을까요?

lhs0610 commented on src/main/kotlin/racingcar/validator/InputArgumentValidators.kt:
```diff
@@ -0,0 +1,27 @@
+    private val validatorMap = mutableMapOf<KClass<*>, InputArgumentValidator>()
+
+    init {
+        addValidator(EntryInput::class, EntryArgumentValidator)
+        addValidator(TotalStepInput::class, TotalStepArgumentValidator)
```
저는 이런 addResolver초기화는 config레벨에서 해주는게 어떨까 싶기는해요. 이 검증기 일급 컬렉션을 쓰기 위해선 항상 두 개의 검증기가 추가되는 것이니까요.

Jinni01 replied:
원래 Configuration 클래스를 별도로 만들어서 명시적 이번 Request change때 지우게 되었는데 이전 리뷰에서 init()메서드를 별도로 만들 필요가 없다 말씀하셨던 거였군요 하하

부활시키고 싱글턴 클래스 스펙으로 다시 만들어보도록 하겠습니다.

lhs0610 commented on src/test/kotlin/racingcar/EntryArgumentValidatorTest.kt:
```diff
@@ -0,0 +1,41 @@
+            validate shouldBe false
+        }
+
+        expect("validate test - racer name length > 5") {
```
테스트에서 5라고 지정을 해뒀는데, 이런 이름의 threshold는 기획에따라 프로덕션 코드(NAME_LIMIT)가 변경될 때 이 부분은 변경되지 않을 취약점이 있을 것 같아요. 

주석도 OOP 패러다임을 따르도록 하면 어떨까요.
그리고 kotest의 최대 장점이 스토리텔링이라는 점도 잘 기억해두면 좋을 것 같네요.

lhs0610 commented on src/test/kotlin/racingcar/RaceHistoryTest.kt:
```diff
@@ -0,0 +1,34 @@
+    context("RaceHistory test") {
+        expect("get winner group test") {
+
+            val step1 = StepHistory(
+                listOf(
+                    StepRacingCarInfo("rc1", 1),
+                    StepRacingCarInfo("rc2", 1)
+                )
+            )
+
+            val step2 = StepHistory(
+                listOf(
+                    StepRacingCarInfo("rc1", 2),
+                    StepRacingCarInfo("rc2", 1)
+                )
+            )
```
context에 있어야 할 부분같지 않나요? 

ExpectSpec으로 context와 expect를 나눠서 사용하는 이유에 대해 고민해보면 좋을 것 같아요.

lhs0610 commented on src/main/kotlin/racingcar/validator/InputArgumentValidator.kt:
```diff
@@ -0,0 +1,5 @@
+package racingcar.validator
+
+interface InputArgumentValidator {
```
메서드가 하나뿐인 인터페이스라면 자바의 FunctionalInterface 애노테이션처럼
인터페이스에 fun을 붙혀줄 수 있습니다. 
```suggestion
fun interface InputArgumentValidator {
```

lhs0610 reviewed (approved):
의진님 피드백 반영 잘 해주셨고, 몇 가지 더 고민해보면 좋을 부분 피드백 남겨놨는데, 중요한 것들은 아니라 다음 미션에서 고민해봐도 좋을 것 같네요. 이만 merge하도록 하겠습니다.

Jinni01 commented:
> 의진님 피드백 반영 잘 해주셨고, 몇 가지 더 고민해보면 좋을 부분 피드백 남겨놨는데, 중요한 것들은 아니라 다음 미션에서 고민해봐도 좋을 것 같네요. 이만 merge하도록 하겠습니다.

해당 내용도 반영해보고 넘어가도록 하겠습니다 ㅎㅎ
